server/game/gameSystems/DeckSearchSystem.ts

```diff
@@ -0,0 +1,284 @@
+import type { AbilityContext } from '../core/ability/AbilityContext';
+import { Card } from '../core/card/Card';
+import { EventName, Location, TargetMode } from '../core/Constants';
+import { DeckSearchEvent } from '../core/event/DeckSearchEvent';
+import { GameEvent } from '../core/event/GameEvent';
+import { GameSystem, IGameSystemProperties } from '../core/gameSystem/GameSystem';
+import { IPlayerTargetSystemProperties, PlayerTargetSystem } from '../core/gameSystem/PlayerTargetSystem';
+import Player from '../core/Player';
+import { shuffleArray } from '../core/utils/Helpers';
+
+type Derivable<T> = T | ((context: AbilityContext) => T);
+
+export interface IDeckSearchProperties extends IPlayerTargetSystemProperties {
+    targetMode?: TargetMode;
+    activePromptTitle?: string;
+    cardsToSearch?: number | ((context: AbilityContext) => number); // Number of cards to search through
+    selectCardCount?: number | ((context: AbilityContext) => number);
+    reveal?: boolean;
+    shuffle?: boolean | ((context: AbilityContext) => boolean);
+    title?: string;
+    immediateEffect?: GameSystem<IGameSystemProperties>; // What should be done with the selected cards?
+    message?: string;
+    uniqueNames?: boolean;
```

Comment:
I think this parameter is probably worth keeping but I think it needs a clearer name